actionmailer/lib/action_mailer/test_helper.rb

```diff
@@ -153,14 +160,29 @@ def assert_enqueued_email_with(mailer, method, params: nil, args: nil, queue: Ac
         params = mailer.instance_variable_get(:@params)
         mailer = mailer.instance_variable_get(:@mailer)
       end
-      args = if args.is_a?(Hash)
-        [mailer.to_s, method.to_s, "deliver_now", params: args, args: []]
-      elsif params.present?
-        [mailer.to_s, method.to_s, "deliver_now", params: params, args: Array(args)]
+
+      expected_args = if args.is_a?(Proc) || params.is_a?(Proc)
+        -> (job_args) {
+          # Ordered from the least to the most heavy checks.
+          return false if [mailer.to_s, method.to_s, "deliver_now"] != job_args[0..2]
+          return false if !params.is_a?(Proc) && params != job_args[3][:params]
+          return false if !args.is_a?(Proc) && Array(args) != job_args[3][:args]
+          return false if args.is_a?(Proc) && !args.(job_args[3][:args])
+          return false if params.is_a?(Proc) && !params.(job_args[3][:params])
+
+          true
+        }
       else
-        [mailer.to_s, method.to_s, "deliver_now", args: Array(args)]
+        if args.is_a?(Hash)
+          [mailer.to_s, method.to_s, "deliver_now", params: args, args: []]
+        elsif params.present?
+          [mailer.to_s, method.to_s, "deliver_now", params: params, args: Array(args)]
+        else
+          [mailer.to_s, method.to_s, "deliver_now", args: Array(args)]
+        end
       end
```

Comment:
This code was a bit hard for me to follow.  I've pushed a commit that unifies the logic.  Let me know what you think.  If you like it, you or I can rebase and squash the commits.